p4-farm-nft/src/lib.rs

```diff
@@ -0,0 +1,2423 @@
+        assert!(
+            now < self.farming_start - ROUND,
+            "must be finalized at last before farm start"
+        );
+        for i in 0..self.farm_deposits.len() {
+            assert_ne!(
+                self.farm_deposits[i], 0,
+                "Deposit for token {} not done",
+                self.farm_tokens[i]
+            )
+        }
+        self.setup_finalized = true;
+    }
+
+    /// Returns expected and received deposits for farmed tokens
+    pub fn finalize_setup_expected(&self) -> (Vec<U128>, Vec<U128>) {
+        //self.assert_owner();
+        let total_rounds = u128::from(round_number(
+            self.farming_start,
+            self.farming_end,
+            self.farming_end,
+        ));
+        log!("rounds: {}", total_rounds);
+        let out = self
+            .farm_token_rates
+            .iter()
+            .map(|rate| farmed_tokens(total_rounds * self.farm_unit_emission, *rate))
+            .collect();
+        (to_U128s(&out), to_U128s(&self.farm_deposits))
+    }
+
+    /*****************
+     * internal methods */
+
+    fn assert_is_active(&self) {
+        assert!(self.setup_finalized, "contract is not setup yet");
+        assert!(self.is_active, "contract is not active");
+    }
+
+    /// Transfers staked(locked) `Cheddar` after successful NFT `unstake` or on `close`.
+    /// If account on `close` - withdraw all staked `amount` Cheddar from user `Vault`
+    /// Else transfer `amount` equals to `Contract.cheddar_rate` and undeclared in args
+    fn transfer_staked_cheddar(&mut self, user: AccountId, amount: Option<Balance>) -> Promise {
+
+        let transfered_amount =  if let Some(_amount) = amount {
+            // total Cheddar staked amound decreased proportionally current user stake
+            U128(_amount)
+        } else {
+            // total Cheddar staked amound decreased to cheddar rate
+            U128(self.cheddar_rate)
+        };
+
+        self.total_cheddar_stake -= transfered_amount.0;
+        log!("@{} unstake Cheddar locked deposit ( {:?} )", user.clone(), transfered_amount);
+        return ext_ft::ext(self.cheddar.clone())
+            .with_attached_deposit(ONE_YOCTO)
+            .with_static_gas(GAS_FOR_FT_TRANSFER)
+            .ft_transfer(user.clone(), transfered_amount, Some("withdraw staked Cheddar".to_string()))
+
+        .then(Self::ext(env::current_account_id())
+            .with_static_gas(GAS_FOR_CALLBACK)
+            .transfer_staked_cheddar_callback(
+                user.clone(),
+                transfered_amount
+            )
+        )
+    }
+
+    /// transfers staked NFT tokens (NFT contract identified by an index in
+    /// self.stake_tokens) back to the user.
+    /// `self.staked_units` must be adjusted in the caller. The callback will fix the
+    /// `self.staked_units` if the transfer will fails.
+    fn transfer_staked_nft_token(&mut self, user: AccountId, nft_contract_i: usize, token_id: TokenId) -> Promise {
```

Comment:
```suggestion
    fn transfer_staked_nft(&mut self, user: AccountId, nft_contract_i: usize, token_id: TokenId) -> Promise {
```

Reply:
NFT is non fungible token, so no need to add extra `_token` suffix.

Reply:
I agree, will update this